nntrainer/dataset/func_data_producer.cpp

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   func_data_producer.cpp
+ * @date   12 July 2021
+ * @brief  This file contains various data producers from a callback
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include <func_data_producer.h>
+
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+
+FuncDataProducer::FuncDataProducer(datagen_cb datagen_cb, void *user_data_) :
+  cb(datagen_cb),
+  user_data(user_data_) {}
+
+FuncDataProducer::~FuncDataProducer() {}
+
+const std::string FuncDataProducer::getType() const {
+  return FuncDataProducer::type;
+}
+
+void FuncDataProducer::setProperty(const std::vector<std::string> &properties) {
+  NNTR_THROW_IF(!properties.empty(), std::invalid_argument)
+    << "properties is not empty, size: " << properties.size();
+}
+
+DataProducer::Gernerator
+FuncDataProducer::finalize(const std::vector<TensorDim> &input_dims,
+                           const std::vector<TensorDim> &label_dims) {
+  NNTR_THROW_IF(!this->cb, std::invalid_argument)
+    << "given callback is nullptr!";
+
+  auto input_data = std::shared_ptr<float *>(new float *[input_dims.size()],
+                                             std::default_delete<float *[]>());
+  auto label_data = std::shared_ptr<float *>(new float *[label_dims.size()],
+                                             std::default_delete<float *[]>());
+
```

Comment:
It looks like we give mem buffer which we allocate. (correct me if I'm wrong) then.. It will be deleted when we destruct dataset. I guess how about the opposite? user gives buffer and we just are using it.